src/Analyzer.Reports/SarifReportWriter.cs

```diff
@@ -0,0 +1,207 @@
+                    if (!this.sarifRun.OriginalUriBaseIds.ContainsKey(UriBaseIdString))
+                    {
+                        this.sarifRun.OriginalUriBaseIds.Add(
+                            UriBaseIdString,
+                            new ArtifactLocation { Uri = new Uri(UriHelper.MakeValidUri(rootPath), UriKind.RelativeOrAbsolute) });
+                    }
+                }
+            }
+        }
+
+        internal Run SarifRun => this.sarifRun;
+
+        private void InitRun()
+        {
+            this.sarifRun = new Run
+            {
+                Tool = new Tool
+                {
+                    Driver = new ToolComponent
+                    {
+                        Name = Constants.ToolName,
+                        FullName = Constants.ToolFullName,
+                        Version = Constants.ToolVersion,
+                        InformationUri = new Uri(Constants.InformationUri),
+                        Organization = Constants.Organization,
+                    }
+                },
+                OriginalUriBaseIds = new Dictionary<string, ArtifactLocation>(),
+            };
+        }
+
+        private ReportingDescriptor ExtractRule(IEvaluation evaluation)
+        {
+            if (!rulesDictionary.TryGetValue(evaluation.RuleId, out ReportingDescriptor rule))
+            {
+                var hasUri = Uri.TryCreate(evaluation.HelpUri, UriKind.RelativeOrAbsolute, out Uri uri);
+                rule = new ReportingDescriptor
+                {
+                    Id = evaluation.RuleId,
+                    // Name = evaluation.RuleId, TBD issue #198
+                    FullDescription = new MultiformatMessageString { Text = AppendPeriod(evaluation.RuleDescription) },
+                    Help = new MultiformatMessageString { Text = AppendPeriod(evaluation.Recommendation) },
+                    HelpUri = hasUri ? uri : null,
+                    MessageStrings = new Dictionary<string, MultiformatMessageString>()
+                    {
+                        { "default", new MultiformatMessageString { Text = AppendPeriod(evaluation.RuleDescription) } }
+                    },
+                    DefaultConfiguration = new ReportingConfiguration { Level = GetLevelFromEvaluation(evaluation) }
+                };
+                rulesDictionary.Add(evaluation.RuleId, rule);
+            }
+            return rule;
+        }
+
+        private void ExtractResult(IEvaluation rootEvaluation, IEvaluation evaluation, string filePath, IList<Result> sarifResults)
+        {
+            foreach (var result in evaluation.Results.Where(r => !r.Passed))
+            {
+                sarifResults.Add(new Result
+                {
+                    RuleId = rootEvaluation.RuleId,
+                    Level = GetLevelFromEvaluation(rootEvaluation),
+                    Message = new Message { Id = "default" }, // should be customized message for each result 
```

Comment:
Should we also create an issue for this?

Reply:
#199 created.